presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java

```diff
@@ -1931,19 +1931,23 @@ private static Timestamp parseTimestamp(String value, Function<String, ZoneId> t
         long epochSecond = LocalDateTime.of(year, month, day, hour, minute, second, 0)
                 .atZone(zoneId)
                 .toEpochSecond();
-        long epochMillis = SECONDS.toMillis(epochSecond);
 
-        Timestamp timestamp;
-        if (epochMillis >= START_OF_MODERN_ERA) {
-            timestamp = new Timestamp(epochMillis);
-        }
-        else {
+        if (epochSecond < START_OF_MODERN_ERA_SECONDS) {
             // slower path, but accurate for historical dates
             GregorianCalendar calendar = new GregorianCalendar(year, month - 1, day, hour, minute, second);
             calendar.setTimeZone(TimeZone.getTimeZone(zoneId));
-            timestamp = new Timestamp(calendar.getTimeInMillis());
+            verify(calendar.getTimeInMillis() % MILLISECONDS_PER_SECOND == 0, "Fractional second when recalculating epochSecond of a historical date: %s", value);
+            epochSecond = calendar.getTimeInMillis() / MILLISECONDS_PER_SECOND;
         }
-        timestamp.setNanos((int) rescale(fractionValue, precision, 9));
+
+        int nanoOfSecond = (int) rescale(fractionValue, precision, 9);
+        if (nanoOfSecond == NANOSECONDS_PER_SECOND) {
+            epochSecond++;
```

Comment:
what about the "before-epoch" values. Are those handled correctly. Do we have tests coverage?

Reply:
The `fractionValue` is positive number, so this is correct for negative timestamp values as well.

Reply:
i do not feel a negative test is needed here, but added anyway

Reply:
Yeah - I had some mind twist and for a moment thought that maybe for before-epoch we should subtract a second. WHich does not make sense at all :)
Thanks for test anyway.